### PR TITLE
Remember that a user has been verified

### PR DESCRIPTION
**Pull request by positivejam**
The way the plugin currently works, if a user has no row in the `wp_usermeta` table for the key `verify-lock`, they are treated as verified. However, it's possible they have no such row in that table because they signed up _before the plugin was installed_. This means you cannot know for certain whether a user has _truly_ been verified, rather than having simply signed up before the plugin was activated.

If this pull request is accepted, it will make it possible to query the database for users whose email addresses have _definitely_ been verified. For example, you could run the query `select user_id, meta_key, meta_value from wp_usermeta where meta_key='verify-lock'` and get results like this:

```
+---------+-------------+----------------------------------+
| user_id | meta_key    | meta_value                       |
+---------+-------------+----------------------------------+
|       2 | verify-lock | unlocked                         |
|       3 | verify-lock | 1ec36168a65ef399ea5b113b86df0dcc |
+---------+-------------+----------------------------------+
```

Consequently, you could know at any given time whether the email addresses in your database have truly been verified or not.

### CODE REVIEW

**Comment by amiut** [comment]
Hi,
It seems nice, but the thing is this way all the users that have registered before are forced to verify their email addresses, this means we need to send the verification link to all of them, we can however consider the previously registered users as `verified`, i mean if the `verify-lock` in their metadata does not exist or equals to `unlocked` they are considered as `verified`

Also if some one needs to make all of their website's users verify their email address, they can send verification email to each one of them with `lock_user()` method and `send_email()` method, i mean something like this:

```
foreach( $user as $users ){
  $verification->lock_user( $user->ID );
  $verification->send_email( $user );
}
```

I mean the verification of previously registered users be optional for site owners

File: email-verify.php [hunk 1]
```diff
@@ -246,10 +250,19 @@ public function authorize_page_id(){
...
 	/**
-	 * Does user needs email validation?
+	 * Return false if user doesn't need validation, stored hash if it does
 	 */
 	public function needs_validation( $user_id ){
-		return boolval( get_user_meta( $user_id, 'verify-lock', true ) );
+		$needs_verification = false;
+		// If the verification metadata doesn't even exist, assume token hasn't been issued. That would indicate that
+		// the user registered at a time when this plugin was not installed/active. Don't require verification in that case.
+		if( metadata_exists( 'user', $user_id, "verify-lock" ) ) {
```
Comment (positivejam) [inline email-verify.php]: If this user registered before the plugin was installed, `metadata_exists()` will return false here, which makes the whole function return false, which means they don't need validation.

**Comment by positivejam** [comment]
> It seems nice, but the thing is this way all the users that have registered before are forced to verify their email addresses

Actually, that's not the case. The way I've written it, existing users do not have to verify. If the meta value does not exist in the database, then the existing user is allowed to login, no problem. See email-verify.php:259:

```php
$needs_verification = false;
if( metadata_exists( 'user', $user_id, "verify-lock" ) ) {
...
}
return $needs_verification;
```

If this user registered before the plugin was installed, `metadata_exists()` will return false here, which makes the whole function return false, which means they don't need validation.

> Also if some one needs to make all of their website's users verify their email address, they can send verification email to each one of them...

Yes, this would certainly work, but it is all or nothing, and the timing with which you do it makes a big difference. Consider this scenario:

1. Admin builds and launches their site.
2. 100 users register.
3. Admin installs Email Verification plugin.
4. 5 more users register, and because the plugin is installed, they verify.
5. Some time passes, and then the admin decides she wants _all_ users to be verified, so she sends verification emails to everyone.

Two problems in this scenario

1. Some users are asked to verify multiple times, which is a bad user-experience.
2. Only way to avoid this is to know exactly when the plugin was activated and filter out users who registered after that time. But what if you didn't record that time? What if you didn't know it would be significant in the future?

### This is the motivation behind my pull request:
If merged, everything should function the same as before, but with one difference: you can easily know, _for certain_, which users are verified, which still need to be verified, and which were never verified because they existed prior to the installation of the plugin.

Whether you decide to merge this PR or not, thank you for your work on this plugin!

**Comment by amiut** [comment]
> If this user registered before the plugin was installed, metadata_exists() will return false here, which makes the whole function return false, which means they don't need validation.

Sorry i dindn't see that line of code, i merged your pull request with `master` branch, please let me know if anything is wrong.

**Comment by positivejam** [comment]
Thanks @amiut!

**Pull request merged**